R/rd-data.R

```diff
@@ -150,16 +151,78 @@ as_data.tag_examples <- function(x, ...,
 
   id_generator <- UniqueId$new()
 
-  html <- purrr::pmap_chr(
+  pieces <- purrr::pmap(
     list(code = code, run = run, show = show),
     format_example_chunk,
     env = child_env(env),
     topic = topic,
     obj_id = id_generator$id
+  ) %>% purrr::flatten()
+
+  is_html <- function(x) {
+    inherits(x, "html")
+  }
+  has_meta <- function(x) {
+    length(attr(x, "knit_meta")) > 0
+  }
+
+  # find html pieces with meta dependencies
+  html_meta <- purrr::map_lgl(
+    pieces,
+    ~ is_html(.x) && has_meta(.x)
+  )
+
+  # no html with dependencies, return one big pre block
+  if (!any(html_meta)) {
+    content <- paste0(
+      c("<pre class='examples'>", unlist(pieces), "</pre>"),
+      collapse = ""
+    )
+
+    out <- list(content = content)
+    return(out)
+  }
+
+  # identify html and break into chunks
+  pre_group <- cumsum(
+    !html_meta | c(FALSE, html_meta[-1] != html_meta[-length(html_meta)])
+  )
+  pre_parts <- split(pieces, pre_group)
+
+  # add surrounding pre tags to non-html blocks
+  out <- purrr::map_if(
+    pre_parts,
+    ~ !is_html(.x[[1]]) && !has_meta(.x[[1]]),
+    ~ paste0(c("<pre class='examples'>", unlist(.x), "</pre>"))
   )
-  paste(html, collapse = "")
+
+  out <- list(content = paste0(unlist(out), collapse = ""))
```

Comment:
What is the type of `out`? I'd prefer a more specific flatten in order to avoid silently propagating data structure problems.